test/cql-pytest/test_prepare.py

```diff
@@ -0,0 +1,154 @@
+    assert prepared.routing_key_indexes == [3, 2, 1, 0]
+
+    prepared = cql.prepare(
+        f"SELECT * FROM {table2} WHERE r1 = ? AND c2 = ? AND p3 = ? AND p1 = ? AND r2 = ? AND p4 = ? AND c1 = ? AND p2 = ? ALLOW FILTERING")
+    assert prepared.routing_key_indexes == [3, 7, 2, 5]
+
+# Test that pk indexes aren't generated when not all partition key columns are restricted using bind variables.
+# In this situation the driver won't be able to calculate the token, so pk indexes should be empty (None).
+def test_composite_pk_no_indexes(cql, table2):
+    prepared = cql.prepare(
+        f"SELECT * FROM {table2}")
+    assert prepared.routing_key_indexes is None
+
+    prepared = cql.prepare(
+        f"SELECT * FROM {table2} WHERE p1 = ? ALLOW FILTERING")
+    assert prepared.routing_key_indexes is None
+
+    prepared = cql.prepare(
+        f"SELECT * FROM {table2} WHERE p1 = ? AND p2 = ? AND p4 = ? ALLOW FILTERING")
+    assert prepared.routing_key_indexes is None
+
+    prepared = cql.prepare(
+        f"SELECT * FROM {table2} WHERE p1 = ? AND p2 = 0 AND p3 = ? AND p4 = ?")
+    assert prepared.routing_key_indexes is None
+
+    prepared = cql.prepare(
+        f"SELECT * FROM {table2} WHERE p1 = ? AND p2 = 0 AND p3 = ? AND p4 = ? AND c1 = ? AND c2 = ?")
+    assert prepared.routing_key_indexes is None
+
+    prepared = cql.prepare(
+        f"SELECT * FROM {table2} WHERE p1 = 0 AND p2 = 1 AND p3 = 2 AND p4 = 3 AND c1 = 5 AND c2 = 5")
+    assert prepared.routing_key_indexes is None
+
+# Test generating pk indexes for a single column partition key using named bind variables.
+def test_single_pk_indexes_named_variables(cql, table1):
+    prepared = cql.prepare(f"SELECT p FROM {table1} WHERE p = :a")
+    assert prepared.routing_key_indexes == [0]
+
+    prepared = cql.prepare(f"SELECT p, c FROM {table1} WHERE c = :a AND p = :b")
+    assert prepared.routing_key_indexes == [1]
+
+# Test generating pk indexes for a composite partition key using named bind variables.
+def test_composite_pk_indexes_named_variables(cql, table2):
+    prepared = cql.prepare(
+        f"SELECT * FROM {table2} WHERE p1 = :a AND p2 = :b AND p3 = :c AND p4 = :d AND c1 = :e AND c2 = :f")
+    assert prepared.routing_key_indexes == [0, 1, 2, 3]
+
+    prepared = cql.prepare(
+        f"SELECT * FROM {table2} WHERE p1 = :f AND p2 = :e AND p3 = :d AND p4 = :c AND c1 = :b AND c2 = :a")
+    assert prepared.routing_key_indexes == [0, 1, 2, 3]
+
+    prepared = cql.prepare(
+        f"SELECT * FROM {table2} WHERE c1 = :a AND c2 = :b AND p1 = :f AND p2 = :e AND p3 = :d AND p4 = :c")
+    assert prepared.routing_key_indexes == [2, 3, 4, 5]
+
+# Test generating pk indexes with named bind variables where the same variable is used multiple times.
+# The test is scylla_only because Scylla treats :x as a single bind variable, but Cassandra thinks
+# that there are two bind variables, both of them named :x.
```

Comment:
How does having two bind variables with the same name work? How are they assigned? Are we sure that this can be considered a Cassandra bug and not intended (albeit strange) behavior? Is there a Cassandra issue about it?

Reply:
More importantly, is Scylla's behavior regarding a duplicate named bind variable different from Cassandra's, regardless of the efficiency of a token-aware driver? I.e., does it return different **results** for the same query? I'll try to write a test for this...

Reply:
I checked, and the reality is even strange than I thought. If you consider the query
```
SELECT * FROM {table1} WHERE p=:x AND c=:x
```
In both Scylla and Cassandra, the following request (in Python driver) works as hoped:
```
cql.execute(stmt, {'x': x})
```
However, if passing unnamed bind values to EXECUTE, Cassandra requires
```
cql.execute(stmt, (x,x)))
```
While Scylla requires
```
cql.execute(stmt, (x,)))
```
Both Scylla and Cassandra complain about the "wrong" choice - trying to pass (x,x) to Scylla says there are two many bind variables, trying to pass (x,) to Cassadra says that "Invalid unset value for column c".

It's not great that we have this difference between Scylla and Cassandra, and I wonder if we can claim that it is a Cassandra bug. I'm not sure at all. Cassandra's behavior does make some sense, especially if you have more than one named bind marker - e.g., consider
```
SELECT * FROM {table1} WHERE p1=:x AND p2=:y AND p3=:x
```
Why is it obvious that binding `(1,2)` will set x to 1? Yes, x is the first bind marker mentioned - but it is also the last ;-)

Reply:
AFAIK the CQL spec doesn't specify what should happen when a named bind variable occurs twice, so we can do whatever works best.
To me Scylla's approach makes more sense - there is a single variable with name `:x`, not two variables that are both named `:x`:.

I'm not sure if there was any discussion about it, I think it was just implemented this way in https://github.com/scylladb/scylladb/pull/10813

Reply:
Opened https://github.com/scylladb/scylladb/issues/15559 to reconsider if this is a Cassandra bug or a Scylla bug.